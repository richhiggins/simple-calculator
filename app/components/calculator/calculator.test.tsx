import React from "react"
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import Calculator from "./calculator"

// number & clear related tests:
describe("render, input & clear related tests:", () => {
  it("it renders the calculator in default state", () => {
    const container = render(<Calculator />)
    expect(container).toMatchSnapshot()
  })

  it("it displays a number", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "6" }))

    expect(screen.getByRole("textbox")).toHaveValue("6")
  })

  it("it displays a concatenated number", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "6" }))
    await user.click(screen.getByRole("button", { name: "3" }))

    expect(screen.getByRole("textbox")).toHaveValue("63")
  })

  it("it prepends a leading decimal point with zero", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "." }))
    expect(screen.getByRole("textbox")).toHaveValue("0.")
  })

  it("it does not allow a repeated decimal point", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "." }))
    await user.click(screen.getByRole("button", { name: "." }))

    expect(screen.getByRole("textbox")).toHaveValue("0.")
  })

  it("it clears the display", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    expect(screen.getByRole("textbox")).toHaveValue("8")

    await user.click(screen.getByRole("button", { name: "C" }))
    expect(screen.getByRole("textbox")).toHaveValue("0")
  })
})

describe("operator related tests:", () => {
  it("it adds two numbers together", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "6" }))
    await user.click(screen.getByRole("button", { name: "+" }))
    await user.click(screen.getByRole("button", { name: "3" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("9")
  })

  it("it subtracts a number from another number", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "6" }))
    await user.click(screen.getByRole("button", { name: "-" }))
    await user.click(screen.getByRole("button", { name: "3" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("3")
  })

  it("it divides a number by another number", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("4")
  })

  it("it multiplies a number by another number", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    await user.click(screen.getByRole("button", { name: "*" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("16")
  })

  it("it adds a number to a previous calculation", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "+" }))
    await user.click(screen.getByRole("button", { name: "1" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("5")
  })

  it("it subtracts a number from a previous calculation", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "-" }))
    await user.click(screen.getByRole("button", { name: "1" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("3")
  })

  it("it divides a number from a previous calculation", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    await user.click(screen.getByRole("button", { name: "+" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("5")
  })

  it("it multiplies a number from a previous calculation", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "8" }))
    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "*" }))
    await user.click(screen.getByRole("button", { name: "3" }))
    await user.click(screen.getByRole("button", { name: "=" }))

    expect(screen.getByRole("textbox")).toHaveValue("12")
  })

  it("it divides zero when no number is input", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))
    expect(screen.getByRole("textbox")).toHaveValue("0")
  })

  it("it multiplies zero when no number is input", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "*" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))
    expect(screen.getByRole("textbox")).toHaveValue("0")
  })

  it("it adds to zero when no number is input", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "+" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))
    expect(screen.getByRole("textbox")).toHaveValue("2")
  })

  it("it subtracts from zero when no number is input", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "-" }))
    await user.click(screen.getByRole("button", { name: "2" }))
    await user.click(screen.getByRole("button", { name: "=" }))
    expect(screen.getByRole("textbox")).toHaveValue("-2")
  })
})

describe("unhappy path/edge case tests:", () => {
  // (could perhaps add more here... coverage looks good though)
  it("it displays zero when operators are pressed without a number input", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "/" }))
    expect(screen.getByRole("textbox")).toHaveValue("0")

    await user.click(screen.getByRole("button", { name: "*" }))
    expect(screen.getByRole("textbox")).toHaveValue("0")
  })

  it("it displays zero when operators are pressed repeatedly without a number input", async () => {
    const user = userEvent.setup()
    render(<Calculator />)

    await user.click(screen.getByRole("button", { name: "+" }))
    await user.click(screen.getByRole("button", { name: "-" }))
    await user.click(screen.getByRole("button", { name: "/" }))
    await user.click(screen.getByRole("button", { name: "*" }))
    await user.click(screen.getByRole("button", { name: "*" }))
    await user.click(screen.getByRole("button", { name: "*" }))
    expect(screen.getByRole("textbox")).toHaveValue("0")
  })
})
