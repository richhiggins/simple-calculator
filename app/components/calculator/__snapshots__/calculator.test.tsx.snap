// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render, input & clear related tests: it renders the calculator in default state 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="calculator"
      >
        <img
          alt="Equal Experts logo"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="40"
          src="/logo.svg"
          style="color: transparent;"
          width="150"
        />
        <input
          aria-live="polite"
          readonly=""
          type="text"
          value="0"
        />
        <div
          class="grid"
        >
          <button
            class="darkGrey"
          >
            C
          </button>
          <button
            class="darkGrey"
            disabled=""
          >
            +/-
          </button>
          <button
            class="darkGrey"
            disabled=""
          >
            %
          </button>
          <button
            class="orange"
          >
            /
          </button>
          <button>
            7
          </button>
          <button>
            8
          </button>
          <button>
            9
          </button>
          <button
            class="orange"
          >
            *
          </button>
          <button>
            4
          </button>
          <button>
            5
          </button>
          <button>
            6
          </button>
          <button
            class="orange"
          >
            -
          </button>
          <button>
            1
          </button>
          <button>
            2
          </button>
          <button>
            3
          </button>
          <button
            class="orange"
          >
            +
          </button>
          <button>
            0
          </button>
          <button>
            .
          </button>
          <button
            class="orange"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="calculator"
    >
      <img
        alt="Equal Experts logo"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="40"
        src="/logo.svg"
        style="color: transparent;"
        width="150"
      />
      <input
        aria-live="polite"
        readonly=""
        type="text"
        value="0"
      />
      <div
        class="grid"
      >
        <button
          class="darkGrey"
        >
          C
        </button>
        <button
          class="darkGrey"
          disabled=""
        >
          +/-
        </button>
        <button
          class="darkGrey"
          disabled=""
        >
          %
        </button>
        <button
          class="orange"
        >
          /
        </button>
        <button>
          7
        </button>
        <button>
          8
        </button>
        <button>
          9
        </button>
        <button
          class="orange"
        >
          *
        </button>
        <button>
          4
        </button>
        <button>
          5
        </button>
        <button>
          6
        </button>
        <button
          class="orange"
        >
          -
        </button>
        <button>
          1
        </button>
        <button>
          2
        </button>
        <button>
          3
        </button>
        <button
          class="orange"
        >
          +
        </button>
        <button>
          0
        </button>
        <button>
          .
        </button>
        <button
          class="orange"
        >
          =
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
